framework/wcf/Extensibility/Transport/AzureQueueStorageClient/AzureQueueStorageClient/AzureQueueStorageOutputChannel.cs

```diff
@@ -0,0 +1,263 @@
+            : base(factory)
+        {
+            //To be implemented
+        }
+
+        #region IOutputChannel_Properties
+        EndpointAddress IOutputChannel.RemoteAddress
+        {
+            get
+            {
+                return this.remoteAddress;
+            }
+        }
+
+        Uri IOutputChannel.Via
+        {
+            get
+            {
+                return this.via;
+            }
+        }
+        #endregion
+
+        public override T GetProperty<T>()
+        {
+            if (typeof(T) == typeof(IOutputChannel))
+            {
+                return (T)(object)this;
+            }
+
+            T messageEncoderProperty = this.encoder.GetProperty<T>();
+            if (messageEncoderProperty != null)
+            {
+                return messageEncoderProperty;
+            }
+
+            return base.GetProperty<T>();
+        }
+
+        /// <summary>
+        /// Open the channel for use. We do not have any blocking work to perform so this is a no-op
+        /// </summary>
+        protected override void OnOpen(TimeSpan timeout)
+        {
+        }
+
+        protected override IAsyncResult OnBeginOpen(TimeSpan timeout, AsyncCallback callback, object state)
+        {
+            return new CompletedAsyncResult(callback, state);
+        }
+
+        protected override void OnEndOpen(IAsyncResult result)
+        {
+            CompletedAsyncResult.End(result);
+        }
+
+
+        #region Socket_Shutdown
+        /// <summary>
+        /// Shutdown ungracefully
+        /// </summary>
+        protected override void OnAbort()
+        {
+
+        }
+
+        /// <summary>
+        /// Shutdown gracefully
+        /// </summary>
+        protected override void OnClose(TimeSpan timeout)
+        {
+
+        }
+
+        protected override IAsyncResult OnBeginClose(TimeSpan timeout, AsyncCallback callback, object state)
+        {
+            this.OnClose(timeout);
+            return new CompletedAsyncResult(callback, state);
+        }
+
+        protected override void OnEndClose(IAsyncResult result)
+        {
+            CompletedAsyncResult.End(result);
+        }
+        #endregion
+
+        #region Send_Synchronous
+        /// <summary>
+        /// Address the Message and serialize it into a byte array.
+        /// </summary>
+        ArraySegment<byte> EncodeMessage(Message message)
+        {
+            try
+            {
+                this.remoteAddress.ApplyTo(message);
+                return encoder.WriteMessage(message, int.MaxValue, parent.BufferManager);
+            }
+            finally
+            {
+                // We have consumed the message by serializing it, so clean up
+                message.Close();
+            }
+        }
+
+        public void Send(Message message)
+        {
+            try
+            {
+                queueClient.SendMessage(message.ToString());
+            }
```

Comment:
Message.ToString looks like it's the correct thing to send, but it isn't. It's a very specific message encoding format which likely won't match what's output by the message encoder. You need to use EncodeMessage. One specific problem you will also have with this is you are missing applying the To address header in the message, which means the service likely won't be able to route the message. 
You are also missing a call to `base.ThrowIfDisposedOrNotOpen();` at the top of this method, and in the finally, when using EncodeMessage you need to return the buffer.
Does AQS support sending binary data? The message encoder could potentially be a binary message encoder unless we restrict it, so you can't presume it's representable as a string.

Reply:
I see it does support sending binary data. You should always use this method to send the data.

Reply:
I just did some more reading, and it looks like AQS uses XML to wrap the message, and it can base64 encode if it's binary but you need to tell it to do that. base64 encoding bloats messages by about 30% so you only want to do that if it is a binary encoded message. [This is the property](https://learn.microsoft.com/en-us/dotnet/api/azure.storage.queues.queueclientoptions.messageencoding?view=azure-dotnet#azure-storage-queues-queueclientoptions-messageencoding) you need to set. My recommendation is to expose the QueueMessageEncoding property on the transport binding element for when someone is using custom binding and have a check if the message encoder binding element is specifically BinaryMessageEncoderBindingElement and require the property to match. On the AzureQueueStorageBinding, have a property called MessageEncoding which exposes an enum (a new one in this project) with the values Binary and Text. In AzureQueueStorageBinding.CreateBindingElements, it checks the value of this property to choose between TextMessageEncodingBindingElement and BinaryMessageEncodingElement. It would also set the QueueMessageEncoding property on the AzureQueueStorageTransportBindingElement based on this value. Take a look at NetHttpBinding.MessageEncoding for an example.  
Doing it this way enables either Text or Binary to be used, and if they are aren't creating a CustomBinding, everything just works with minimal effort. If they create a custom binding and providing the message encode themselves, they can override the format in the transport to match.

Reply:
Followed the example and sending binary data to Send and exposed a QueueMessageEncoding  property.